lib/streamlit/elements/widgets/selectbox.py

```diff
@@ -54,25 +53,53 @@
     from streamlit.delta_generator import DeltaGenerator
 
 
-@dataclass
 class SelectboxSerde(Generic[T]):
     options: Sequence[T]
-    index: int | None
+    formatted_option_to_option_mapping: dict[str, T]
+    default_option_index: int | None
+    format_func: Callable[[Any], str]
+    accept_new_options: bool
 
-    def serialize(self, v: object) -> int | None:
+    def __init__(
+        self,
+        options: Sequence[T],
+        formatted_option_to_option_mapping: dict[str, T],
+        default_option_index: int | None = None,
+        format_func: Callable[[Any], str] = str,
+        accept_new_options: bool = False,
+    ):
+        self.options = options
+        self.formatted_option_to_option_mapping = formatted_option_to_option_mapping
+        self.default_option_index = default_option_index
+        self.format_func = format_func
+        self.accept_new_options = accept_new_options
+
+    def serialize(self, v: T | str | None) -> str | None:
         if v is None:
             return None
         if len(self.options) == 0:
-            return 0
-        return index_(self.options, v)
+            return ""
+        return self.format_func(v)
```

Comment:
Why does it need `format_func`, wouldn't it be enough to use `formatted_option_to_option_mapping`. 
Storing `format_func` as part of the series across reruns could potentially lead to some unexpected errors, but not 100% sure.

Reply:
The `format_func` is not stored part of the series but is hold in the `SerializerDeserializer` (Serde). The main reason for this is that when `serialize` is called, e.g. when the widget value changed [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/elements/widgets/selectbox.py#L353), the option has to be formatted correctly before being sent to the frontend.
The `formatted_option_to_option_mapping` structure is introduced to efficiently map from an incoming selected option (frontend -> backend) in the deserializer to the generic option. We would lose this efficient lookup with a `option_to_formatted_option` map.  

Could you elaborate a little bit on your concern? I am happy to make a change here if you think that it's risky but I don't understand it yet 🙂

Reply:
> We would lose this efficient lookup with a option_to_formatted_option map.

but we could have both mappings?

Reply:
Ahh ok you mean because we pass the `serde` to the `register_widget` function where it later is accessed in the session state later. Yeah, I agree this feels iffy. I will update it to use two mappings 👍 Thanks for walking me through your thoughts

Reply:
yep 👍 this aspect could open a door for some unexpected issues

Reply:
I have updated the two Serdes to not take the `format_func` as an argument anymore. Instead, it now takes the list of formatted options, which can be used to map the `options` index to the correct formatted option